src/EffectWatcher.re

```diff
@@ -0,0 +1,35 @@
+type context = {revEffects: ref(list(Effect.t))};
+
+module Internal = {
+  let unhandledEffectEvent = Event.create();
+  let currentContext: ref(option(context)) = ref(None);
+};
+
+let onUnhandledEffect = handler => {
+  Event.add(handler, Internal.unhandledEffectEvent);
+};
+
+let notifyEffect = effect => {
+  switch (Internal.currentContext^) {
+  | None => Event.dispatch(effect, Internal.unhandledEffectEvent)
+  | Some(context) => context.revEffects := [effect, ...context.revEffects^]
+  };
+};
+
+let runAndTrackEffects = f => {
+  let context =
+    switch (Internal.currentContext^) {
+    | None => {revEffects: ref([])}
+    | Some(context) => context
+    };
+
+  Internal.currentContext := Some(context);
+  let ret = f();
```

Comment:
I assume this is run here because it produces side-effects. But as I don't think that's all that obvious, it would be nice to have a comment explaining that. Or a more descriptive function name, but I don't have any suggestions myself for it.

Reply:
Good point - I'm holding off on the effect / effectwatcher PR, just to focus on the context & removing the 'synchronization' in Onivim 2. Will bring that in separately.